### PR TITLE
cmd/geth, console: support interrupting the js console

### LINKED ISSUE #23344: Geth console unresponsive when running js statement
I would expect to be able to stop a statement (e.g. a loop) with `Ctrl+C` or `Ctrl+D`, but it just carries right along.

### PR DESCRIPTION
**Pull request by holiman**
This PR changes the behavior of Ctrl-C (i.e. SIGINT) when the console is active.
Previously, this signal was ignored during JS execution, so it was not possible
to get out of infinite loops in the console. With this change, Ctrl-C now
interrupts JS.

Fixes #23344

### CODE REVIEW

**Comment by fjl** [comment]
I have some ideas for this, please don't merge it yet.

**Comment by s1na** [comment]
@fjl was curious what your idea was regarding this PR

**Comment by fjl** [comment]
I tried to implement the proper semantics where geth doesn't shut down on Ctrl-C. It's a bit complicated though.
My idea was to refactor the [geth interrupt handler](https://github.com/ethereum/go-ethereum/blob/85064ed09b7243647ffd8c31a93c2abdbc3850d2/cmd/utils/cmd.go) so that it wouldn't be active in console mode.

**Comment by s1na** [comment]
I was getting a crash immediately during start. I couldn't figure out where it came from but after a rebase against master the crash is gone

**Comment by s1na** [comment]
> My idea was refactoring the geth interrupt handler so that it wouldn't be active in console mode.

I tried your idea [here](https://github.com/s1na/go-ethereum/commit/94a6936f56839cb64a039c686fe0d7bf2370577d). It works. I thought sigterm should still cause shutdown. Do you agree with that? if so I need to change the code a bit because right now if first a sigint arrives and we're in console mode any future sigterm is ignored.

**Comment by s1na** [comment]
@holiman can you please take a look at my changes whenever you get a chance and tell me if its okay for me to push them to your branch?

**Comment by holiman** [comment]
I find it a bit hard to follow, can't compare across forks, since you rebased mine before you committed yours. Do you think you can push the rebased to my repo for easier comparison?

**Comment by s1na** [comment]
Yes squashed: https://github.com/s1na/go-ethereum/commit/0e99d3ffae557ffdf51ac93bbd3f5f4661eaa8f1

**Comment by holiman** [comment]
LGTM, push it here why dontcha?

**Comment by holiman** [comment]
Yup, with your fixes, this now works for me:
```
[user@work go-ethereum]$ ./build/bin/geth --dev console
...
Welcome to the Geth JavaScript console!

instance: Geth/v1.10.14-unstable-46891a6b-20211203/linux-amd64/go1.17.3
coinbase: 0xc9493a99d64c05dbfcf6635a52aa7a990f0250bc
at block: 0 (Thu Jan 01 1970 01:00:00 GMT+0100 (CET))
 datadir: 
 modules: admin:1.0 clique:1.0 debug:1.0 eth:1.0 miner:1.0 net:1.0 personal:1.0 rpc:1.0 txpool:1.0 web3:1.0

To exit, press ctrl-d or type exit
> while(true){}
^C> aborted by caller at <eval>:1:1(1)

```

**Comment by fjl** [comment]
I will re-check and merge later today!

**Comment by fjl** [comment]
I've tested this extensively and decided to rework the interrupt logic a bit more to
really handle all the cases. In some of the cases listed below, the JS interrupt did not
work or caused an unclean shutdown of the node.


```
./geth console --dev

> partialInput^
# when SIGTERM received, geth should exit
# when SIGINT received, it should not exit
# when Ctrl-C, it should clear the line
```


```
./geth console --dev

> while(true) { admin.sleep(1); console.log("sleeping"); }
sleeping
sleeping
^
# when SIGTERM received, geth should exit
# when SIGINT received or Ctrl-C, it should return to JS prompt
```


```
TERM=dumb ./geth console --dev

> partialInput^
# special case: when Ctrl-C/SIGINT, geth should exit
```


```
TERM=dumb ./geth console --dev

> while(true) { admin.sleep(1); console.log("sleeping"); }
sleeping
sleeping
^
# when SIGTERM received, geth should exit
# when SIGINT received or Ctrl-C, it should return to JS prompt
```


```
./geth console --dev --exec 'while(true) { admin.sleep(1); console.log("sleeping"); }'
sleeping
sleeping
^
# SIGTERM/SIGINT/Ctrl-C should interrupt JS and shut down the node
```


```
echo 'while(true) { admin.sleep(1); console.log("sleeping"); }' > loop1.js
./geth --dev js loop1.js
sleeping
sleeping
^
# SIGTERM/SIGINT/Ctrl-C should interrupt JS and shut down the node
```

**Comment by holiman** [comment]
LGTM. While testing this, I had some issues with the history, things didn't seem to be added to it. Then I cleared out the history (there were `1923` lines in it), and tried to repro it, but suddenly things started to work after that, so not sure what it was.

**Comment by fjl** [comment]
I need to recheck my change to the history write. Maybe it doesn't work properly now.

**Comment by fjl** [comment]
I checked the history behavior and it seems to work.
